releasing/plugin-commands-deploy/test/deploy.test.ts

```diff
@@ -77,6 +97,513 @@ test('deploy', async () => {
+  preparePackages([
+    {
+      name: 'project-0',
+      version: '0.0.0',
+      private: true,
+      dependencies: {
+        'project-1': 'workspace:*',
+      },
+    },
+    {
+      name: 'project-1',
+      version: '0.0.0',
+      private: true,
+      dependencies: {
+        'project-0': 'workspace:*',
+      },
+    },
+  ])
+
+  const {
+    allProjects,
+    allProjectsGraph,
+    selectedProjectsGraph,
+  } = await filterPackagesFromDir(process.cwd(), [{ namePattern: 'project-0' }])
+
+  await install.handler({
+    ...DEFAULT_OPTS,
+    allProjects,
+    allProjectsGraph,
+    selectedProjectsGraph: allProjectsGraph,
+    dir: process.cwd(),
+    recursive: true,
+    lockfileDir: process.cwd(),
+    workspaceDir: process.cwd(),
+  })
+  expect(fs.existsSync('pnpm-lock.yaml')).toBeTruthy()
+
+  await deploy.handler({
+    ...DEFAULT_OPTS,
+    allProjects,
+    dir: process.cwd(),
+    recursive: true,
+    selectedProjectsGraph,
+    sharedWorkspaceLockfile: true,
+    lockfileDir: process.cwd(),
+    workspaceDir: process.cwd(),
+  }, ['deploy'])
+
+  const project = assertProject(path.resolve('deploy'))
+  project.has('project-1')
+
+  const lockfile = project.readLockfile()
+  expect(lockfile.importers).toStrictEqual({
+    '.': {
+      dependencies: {
+        'project-1': {
+          version: expect.stringContaining('project-1'),
+          specifier: expect.stringContaining('file:'),
+        },
+      },
+    },
+  } as LockfileFile['importers'])
+
+  const manifest = readPackageJson('deploy') as ProjectManifest
+  expect(manifest).toStrictEqual({
+    name: 'project-0',
+    version: '0.0.0',
+    private: true,
+    dependencies: {
+      'project-1': expect.stringContaining('project-1'),
+    },
+    devDependencies: {},
+    optionalDependencies: {},
+  } as ProjectManifest)
+
+  const project1Name = fs.readdirSync('deploy/node_modules/.pnpm').find(name => name.includes('project-1@'))
+  expect(project1Name).toBeDefined()
+  expect(fs.readdirSync(`deploy/node_modules/.pnpm/${project1Name}/node_modules`).sort()).toStrictEqual(['project-0', 'project-1'])
+  expect(fs.realpathSync(`deploy/node_modules/.pnpm/${project1Name}/node_modules/project-0`)).toBe(path.resolve('deploy'))
+  expect(fs.realpathSync('deploy/node_modules/project-1')).toBe(path.resolve(`deploy/node_modules/.pnpm/${project1Name}/node_modules/project-1`))
+})
+
+test('deploy with a shared lockfile should correctly handle package that depends on itself', async () => {
+  preparePackages([
+    {
+      name: 'project-0',
+      version: '0.0.0',
+      private: true,
+      dependencies: {
+        'project-0': 'workspace:*',
+        // NOTE: there is an inexplicable bug in which 'renamed-workspace' and 'renamed-linked' cannot exist simultaneously
+        // TODO: fix this bug
+        'renamed-workspace': 'workspace:project-0@*',
+        // 'renamed-linked': 'link:.',
```

Comment:
I cannot reproduce this bug with `pd`. I don't know how to fix this.

Reply:
Fixed by https://github.com/pnpm/pnpm/pull/8828/commits/ba79dff7278976e27c47947ccdad19fde9d5d513